File: core/src/main/java/nl/t64/game/rpg/components/party/PartyContainer.java
package nl.t64.game.rpg.components.party;

import nl.t64.game.rpg.constants.Constant;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;


public class PartyContainer {

    public static final int MAXIMUM = 6;
    private final Map<String, HeroItem> party;

    public PartyContainer() {
        this.party = new LinkedHashMap<>(MAXIMUM);
    }

    public HeroItem getHero(int index) {
        return getAllHeroes().get(index);
    }

    public int getIndex(HeroItem hero) {
        return getAllHeroes().indexOf(hero);
    }

    public List<HeroItem> getAllHeroes() {
        return List.copyOf(party.values());
    }

    public HeroItem getPreviousHero(HeroItem hero) {
        if (isHeroFirst(hero)) {
            return getLastHero();
        } else {
            return getHero(getIndex(hero) - 1);
        }
    }

    public HeroItem getNextHero(HeroItem hero) {
        if (isHeroLast(hero)) {
            return getFirstHero();
        } else {
            return getHero(getIndex(hero) + 1);
        }
    }

    public void addHero(HeroItem hero) {
        if (isFull()) {
            throw new IllegalStateException("Party is full.");
        }
        party.put(hero.id, hero);
    }

    public void removeHero(String heroId) {
        if (isPlayer(heroId)) {
            throw new IllegalArgumentException("Cannot remove player from party.");
        }
        party.remove(heroId);
    }

    public boolean isFull() {
        return getSize() >= MAXIMUM;
    }

    public boolean containsExactlyEqualTo(HeroItem candidateHeroObject) {
        if (candidateHeroObject == null) {
            return false;
        }
        final HeroItem possibleFoundHero = getHero(candidateHeroObject.id);
        return possibleFoundHero != null && possibleFoundHero.equals(candidateHeroObject);
    }

    public boolean isHeroLast(HeroItem hero) {
        return getIndex(hero) == getLastIndex();
    }

    private boolean isHeroFirst(HeroItem hero) {
        return getIndex(hero) == 0;
    }

    public int getSize() {
        return party.size();
    }

    HeroItem getHero(String heroId) {
        return party.get(heroId);
    }

    boolean contains(String heroId) {
        return party.containsKey(heroId);
    }

    int getSumOfSkill(SkillItemId skillItemId) {
        return getAllHeroes().stream()
                             .mapToInt(heroItem -> heroItem.getCalculatedTotalSkillOf(skillItemId))
                             .sum();
    }

    private HeroItem getFirstHero() {
        return getHero(0);
    }

    private HeroItem getLastHero() {
        return getHero(getLastIndex());
    }

    private int getLastIndex() {
        return getSize() - 1;
    }

    private boolean isPlayer(String heroId) {
        return heroId.equals(Constant.PLAYER_ID);
    }

}

File: core/src/nl/t64/game/rpg/components/party/PartyContainer.kt
package nl.t64.game.rpg.components.party

import nl.t64.game.rpg.components.party.skills.SkillItemId
import nl.t64.game.rpg.constants.Constant


class PartyContainer {

    companion object {
        const val MAXIMUM = 6
    }

    private val party: MutableMap<String, HeroItem> = LinkedHashMap(MAXIMUM)
    private val firstHero: HeroItem get() = getHero(0)
    private val lastHero: HeroItem get() = getHero(lastIndex)
    private val lastIndex: Int get() = size - 1
    val isFull: Boolean get() = size >= MAXIMUM
    val size: Int get() = party.size

    fun gainXp(amount: Int, levelUpMessage: StringBuilder) {
        getAllHeroesAlive().forEach { it.gainXp(amount, levelUpMessage) }
    }

    fun getPreviousHero(hero: HeroItem): HeroItem {
        return when {
            isHeroFirst(hero) -> lastHero
            else -> getHero(getIndex(hero) - 1)
        }
    }

    fun getNextHero(hero: HeroItem): HeroItem {
        return when {
            isHeroLast(hero) -> firstHero
            else -> getHero(getIndex(hero) + 1)
        }
    }

    fun addHero(hero: HeroItem) {
        check(!isFull) { "Party is full." }
        party[hero.id] = hero
        hero.hasBeenRecruited = true
    }

    fun removeHero(heroId: String) {
        require(!isPlayer(heroId)) { "Cannot remove player from party." }
        party.remove(heroId)
    }

    fun isPlayer(heroId: String): Boolean {
        return heroId == Constant.PLAYER_ID
    }

    fun containsExactlyEqualTo(candidateHeroObject: HeroItem): Boolean {
        return getPossibleHero(candidateHeroObject.id) == candidateHeroObject
    }

    fun isHeroLast(hero: HeroItem): Boolean {
        return getIndex(hero) == lastIndex
    }

    private fun isHeroFirst(hero: HeroItem): Boolean {
        return getIndex(hero) == 0
    }

    fun getSumOfSkill(skillItemId: SkillItemId): Int {
        return getAllCalculatedTotalSkillsOf(skillItemId).sum()
    }

    fun getHeroWithHighestSkill(skillItemId: SkillItemId): HeroItem {
        return getAllHeroesAlive().maxByOrNull { it.getCalculatedTotalSkillOf(skillItemId) } ?: firstHero
    }

    fun hasEnoughOfSkill(skillItemId: SkillItemId, rank: Int): Boolean {
        return getBestSkillLevel(skillItemId) >= rank
    }

    fun contains(heroId: String): Boolean {
        return party.containsKey(heroId)
    }

    fun getHero(index: Int): HeroItem {
        return getAllHeroes()[index]
    }

    fun getIndex(hero: HeroItem): Int {
        return getAllHeroes().indexOf(hero)
    }

    fun getAllHeroesAlive(): List<HeroItem> = getAllHeroes().filter { it.isAlive }

    fun getAllHeroes(): List<HeroItem> = ArrayList(party.values)

    fun getCertainHero(heroId: String): HeroItem {
        return getPossibleHero(heroId)!!
    }

    private fun getPossibleHero(heroId: String): HeroItem? {
        return party[heroId]
    }

    private fun getBestSkillLevel(skillItemId: SkillItemId): Int {
        return getAllCalculatedTotalSkillsOf(skillItemId).maxOfOrNull { it } ?: 0
    }

    private fun getAllCalculatedTotalSkillsOf(skillItemId: SkillItemId): List<Int> {
        return getAllHeroesAlive().map { it.getCalculatedTotalSkillOf(skillItemId) }
    }

}
